concordium-consensus/src/Concordium/KonsensusV1/Consensus/Finality.hs

```diff
@@ -89,13 +90,88 @@ processCertifiedBlock cb@CertifiedBlock{..}
         alreadyStored <- LowLevel.memberBlock (getHash cbQuorumBlock)
         unless alreadyStored a
 
+-- | Result of catching up a 'FinalizationEntry'.
+data CatchupFinalizationEntryResult
+    = -- | The 'FinalizationEntry' is valid and consensus
+      --  processed it.
+      CFERSuccess
+    | -- | The proposed finalized block is inconsistent
+      --  with the finalized block being pointed to in
+      --  the 'FinalizationEntry'.
+      CFERInconsistent
+    | -- | The finalization entry is invalid.
+      CFERInvalid
+    | -- | The finalization entry pointed to a block that was not
+      --  alive.
+      CFERNotAlive
+    | -- | The baking committee could not be looked up for the 'Epoch'
+      --  of the finalized 'QuorumCertificate'.
+      CFERUnknownBakers
+
+-- | Receive a 'FinalizationEntry' as part of catch-up.
+--  This function checks whether the finalization entry is
+--  consistent with the block indicated by @feFinalizedQuorumCertificate@ of the
+--  finalization entry and that block is alive and non finalized.
+--  If the finalization entry can be verified then it is being processed.
+catchupFinalizationEntry ::
+    ( MonadState (SkovData (MPV m)) m,
+      TimeMonad m,
+      MonadIO m,
+      LowLevel.MonadTreeStateStore m,
+      BlockStateStorage m,
+      GSTypes.BlockState m ~ PBS.HashedPersistentBlockState (MPV m),
+      MonadThrow m,
+      MonadConsensusEvent m,
+      MonadLogger m,
+      IsConsensusV1 (MPV m)
+    ) =>
+    -- | The finalization entry received.
+    FinalizationEntry ->
+    -- | Whether the finalization entry was successfully processed.
+    m CatchupFinalizationEntryResult
+catchupFinalizationEntry finEntry = do
+    let finQC = feFinalizedQuorumCertificate finEntry
+    let finBlockHash = qcBlock finQC
+    gets (getLiveBlock finBlockHash) >>= \case
+        Just block ->
+            checkConsistency finQC block $
```

Comment:
```suggestion
        Just block
            | blockRound block /= qcRound qc || blockEpoch block /= qcEpoch qc -> return CFERInconsistent
            | otherwise ->
```